subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/GroovyJavaJointIncrementalCompilationIntegrationTest.groovy

```diff
@@ -51,32 +74,28 @@ class GroovyJavaJointIncrementalCompilationIntegrationTest extends AbstractJavaG
         run "compileGroovy", "--info"
-        executedAndNotSkipped(":compileGroovy")
-        outputContains(firstRecompilationReason)
+        upToDateOrMessage(firstBuildMessage)
+        outputContains(firstBuildMessage)
 
-        when:
+        when: 'second build'
         applyFileSet(secondChange)
         run "compileGroovy", "--info"
 
         then:
-        if (secondRecompilationReason == 'UP-TO-DATE') {
-            skipped(":compileGroovy")
-        } else {
-            executedAndNotSkipped(":compileGroovy")
-        }
-        outputContains(secondRecompilationReason)
+        upToDateOrMessage(secondBuildMessage)
+        outputContains(secondBuildMessage)
 
         where:
-        initialSet             | firstChange                         | firstRecompilationReason                 | secondChange                        | secondRecompilationReason
-        ['G.groovy']           | ['G.groovy', 'J.java']              | 'Groovy-Java joint compilation detected' | ['G.groovy', 'J.java']              | 'UP-TO-DATE'
-        ['J.java']             | ['G.groovy', 'J.java']              | 'no source class mapping file found'     | ['G.groovy', 'J.java.changed']      | 'no source class mapping file found'
-        ['G.groovy', 'J.java'] | ['G.groovy', 'J.java.changed']      | 'no source class mapping file found'     | ['G.groovy', 'J.java.changed']      | 'UP-TO-DATE'
-        ['G.groovy', 'J.java'] | ['G.groovy.changed', 'J.java']      | 'no source class mapping file found'     | ['G.groovy.changed', 'J.java']      | 'UP-TO-DATE'
-        ['G.groovy', 'J.java'] | ['G.groovy']                        | 'no source class mapping file found'     | ['G.groovy', 'G2.groovy']           | 'Incremental compilation of '
-        ['G.groovy', 'J.java'] | ['J.java']                          | 'no source class mapping file found'     | ['J.java', 'J2.java']               | 'no source class mapping file found'
-        ['G.groovy', 'J.java'] | ['G.groovy', 'J.java', 'J2.java']   | 'no source class mapping file found'     | ['G.groovy', 'G2.groovy']           | 'no source class mapping file found'
-        ['G.groovy', 'J.java'] | ['G.groovy', 'J.java', 'G2.groovy'] | 'no source class mapping file found'     | ['G.groovy', 'J.java', 'G2.groovy'] | 'UP-TO-DATE'
+        initialSet                            | firstChange                                   | firstBuildMessage                                        | secondChange                                  | secondBuildMessage
+        ['G.groovy']                          | ['G.groovy', 'J.java']                        | 'Groovy-Java joint compilation detected'                 | ['G.groovy', 'J.java']                        | 'UP-TO-DATE'
+        ['J.java']                            | ['G.groovy', 'J.java']                        | 'no source class mapping file found'                     | ['G.groovy', 'J.java.changed']                | 'Groovy-Java joint compilation detected'
```

Comment:
I think we should have a better rebuild clause than `no source class mapping file found`. When can this message appear?

IIUC, this can happen only when the Groovy sources do not produce any classes. I think you said that this is not possible. So we end up having only non-Groovy source files.

So maybe something like this would be more appropriate:
```
Last compilation was unable to determine the relationship between sources and class files
```

Reply:
Two  cases:

1. Only non-Groovy files in previous build.
2. The mapping file are accidentally deleted.

Changed to `unable to get source-classes mapping relationship from last compilation`